velox/exec/tests/TaskTest.cpp

```diff
@@ -1373,77 +1373,169 @@ DEBUG_ONLY_TEST_F(TaskTest, findPeerOperators) {
+  ASSERT_EQ(task_->numTotalDrivers(), 1);
+  ASSERT_EQ(task_->numFinishedDrivers(), 0);
+  ASSERT_EQ(task_->numRunningDrivers(), 0);
 
-  taskThread.join();
+  Task::resume(task_->shared_from_this());
+
+  ASSERT_EQ(task_->numTotalDrivers(), 1);
+  ASSERT_EQ(task_->numFinishedDrivers(), 1);
+  ASSERT_EQ(task_->numRunningDrivers(), 0);
+
+  taskThread_.join();
+}
+
+DEBUG_ONLY_TEST_F(TaskPauseTest, resumeFuture) {
+  // Test for trivial wait on Task::pauseRequested future.
+  testPause();
+  folly::EventCount taskResumeAllowedWait;
+  std::atomic<bool> taskResumeAllowed{false};
+  std::thread observeThread([&]() {
+    ContinueFuture future = ContinueFuture::makeEmpty();
+    const bool requested = task_->pauseRequested(&future);
+    ASSERT_TRUE(requested);
+    taskResumeAllowed = true;
+    taskResumeAllowedWait.notifyAll();
+    future.wait();
+    ASSERT_FALSE(task_->pauseRequested());
+  });
+
+  taskResumeAllowedWait.await([&]() { return taskResumeAllowed.load(); });
+  std::this_thread::sleep_for(std::chrono::milliseconds(100)); // NOLINT
+  Task::resume(task_->shared_from_this());
+
+  taskThread_.join();
+  observeThread.join();
+
+  ASSERT_EQ(task_->numTotalDrivers(), 1);
+  ASSERT_EQ(task_->numFinishedDrivers(), 1);
+  ASSERT_EQ(task_->numRunningDrivers(), 0);
+}
+
+DEBUG_ONLY_TEST_F(TaskPauseTest, resumeFutureAfterTaskTerminated) {
+  testPause();
+  folly::EventCount taskCancelAllowedWait;
+  std::atomic<bool> taskCancelAllowed{false};
+  folly::EventCount taskCanceledWait;
+  std::atomic<bool> taskCanceled{false};
+  folly::EventCount taskResumeAllowedWait;
+  std::atomic<bool> taskResumeAllowed{false};
+  folly::EventCount taskResumedWait;
+  std::atomic<bool> taskResumed{false};
+  std::thread observeThread([&]() {
+    ContinueFuture future = ContinueFuture::makeEmpty();
+    const bool requested = task_->pauseRequested(&future);
+    ASSERT_TRUE(requested);
+    taskCancelAllowed = true;
+    taskCancelAllowedWait.notifyAll();
+    taskCanceledWait.await([&]() { return taskCanceled.load(); });
+    std::this_thread::sleep_for(std::chrono::milliseconds(100)); // NOLINT
+    ASSERT_FALSE(future.isReady());
+    taskResumeAllowed = true;
+    taskResumeAllowedWait.notifyAll();
+    taskResumedWait.await([&]() { return taskResumed.load(); });
+    std::this_thread::sleep_for(std::chrono::milliseconds(100)); // NOLINT
+    ASSERT_TRUE(future.isReady());
+    future.wait();
+    ASSERT_FALSE(task_->pauseRequested());
+  });
+  taskCancelAllowedWait.await([&]() { return taskCancelAllowed.load(); });
+  std::this_thread::sleep_for(std::chrono::milliseconds(100)); // NOLINT
+  ASSERT_EQ(task_->numTotalDrivers(), 1);
+  ASSERT_EQ(task_->numFinishedDrivers(), 0);
+  ASSERT_EQ(task_->numRunningDrivers(), 1);
+  std::vector<Driver*> drivers{};
+  task_->requestCancel().wait();
+  ASSERT_EQ(task_->numTotalDrivers(), 1);
+  ASSERT_EQ(task_->numFinishedDrivers(), 0);
+  ASSERT_EQ(task_->numRunningDrivers(), 0);
+  taskCanceled = true;
```

Comment:
I don't think you need a separate thread. You can just check all in the test thread.